### PR TITLE
foxglove-websocket: add version 0.0.1

### PR DESCRIPTION
**Pull request by jtbandes**
Hello 👋

I'm the author of this package, `foxglove-websocket`, which is part of a larger project at [foxglove/ws-protocol](https://github.com/foxglove/ws-protocol) ([read more here](https://foxglove.dev/blog/announcing-the-foxglove-websocket-protocol)). We would like the C++ server library to be available on ConanCenter. This is the first version of the package. It's currently a header-only library, with dependencies on `nlohmann_json` and `websocketpp`.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.
  - Built in a `conanio/clang10` Docker container using `-s compiler.cppstd=17 -s compiler.libcxx=libc++`.

### CODE REVIEW

File: recipes/foxglove-websocket/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+class FoxgloveWebSocketConan(ConanFile):
+    name = "foxglove-websocket"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/foxglove/ws-protocol"
+    description = "A C++ server implementation of the Foxglove WebSocket Protocol"
+    license = "MIT"
+    topics = ("foxglove", "websocket")
+
+    settings = ("os", "compiler", "build_type", "arch")
+    requires = ("nlohmann_json/[^3.10.4]", "websocketpp/[^0.8.2]")
```
Comment (prince-chrismc) [inline recipes/foxglove-websocket/all/conanfile.py:14]: We avoid version ranges in CCI... it was not documented so I tried to note done the motivation for this in a new pr https://github.com/conan-io/conan-center-index/pull/8831

 ```suggestion
    requires = ("nlohmann_json/3.10.4", "websocketpp/0.8.2")
```

File: recipes/foxglove-websocket/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    generators = "cmake"
+
+    _source_root = "source_root"
+    _source_package_path = os.path.join(_source_root, "cpp", "foxglove-websocket")
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version], strip_root=True, destination=self._source_root)
+
+    def validate(self):
+        tools.check_min_cppstd(self, "17")
```
Comment (prince-chrismc) [inline recipes/foxglove-websocket/all/conanfile.py:24]: https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#why-is-a-toolscheck_min_cppstd-call-not-enough

Please see above and add the missing `if` 🙏

**Review by prince-chrismc — COMMENTED** [review]
Looks really freaking good! Just two small details to keep it inline with CCI

Having used websocketpp in the past this looks like a welcomed improvement!

File: recipes/foxglove-websocket/all/conanfile.py (hunk 2 above)
Comment (jtbandes) [reply recipes/foxglove-websocket/all/conanfile.py]: Thanks, updated

**Comment by jtbandes** [comment]
Hm, I wonder if this new error is due to the `tools.check_min_cppstd` not running?

**Comment by SSE4** [comment]
> Hm, I wonder if this new error is due to the `tools.check_min_cppstd` not running? Any advice @prince-chrismc?

yes, I think so. it tries to compile with old standard, but as I can see, it needs C++17 at very least, right?

**Comment by jtbandes** [comment]
Yes, that was the purpose of adding `check_min_cppstd`. How else can I achieve this?

File: recipes/foxglove-websocket/all/test_package/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package CXX)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_executable(test_package test_package.cpp)
+target_link_libraries(test_package ${CONAN_LIBS})
```
Comment (prince-chrismc) [inline recipes/foxglove-websocket/all/test_package/CMakeLists.txt:8]: Looking quickly you'll need to configure C++17 here ... here's an example https://github.com/conan-io/conan-center-index/blob/41d4fbb8c218f9d938ccb8b9d888bff52002e62e/recipes/libsolace/all/test_package/CMakeLists.txt#L12

I'll take a better look tomorrow of theres anything else I can find

Comment (jtbandes) [reply recipes/foxglove-websocket/all/test_package/CMakeLists.txt:9]: I think that helped for setting the language standard but still running into problems with the standard library version.

**Comment by prince-chrismc** [comment]
Yuck. My system was out of date... Sadly everything checks out on my setup 🤔 

`CXX=g++-7 CC=gcc-7 conan create all 0.0.1@ -s compiler.version=7` works just fine
I am running a newer gcc 7.5 instead of 7.2 in the docker images

Maybe try bumping the min compiler to 8?

**Review by prince-chrismc — DISMISSED** [review]
🚀

File: recipes/foxglove-websocket/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    def validate(self):
+        if self.settings.compiler.get_safe("cppstd"):
+            tools.check_min_cppstd(self, "17")
+        if (self.settings.compiler == "gcc" or self.settings.compiler == "clang") and tools.Version(self.settings.compiler.version) <= 8:
+            raise ConanInvalidConfiguration("Compiler version is not supported, c++17 support is required")
+        if self.settings.compiler == "Visual Studio" and tools.Version(self.settings.compiler.version) <= "16.8":
+            raise ConanInvalidConfiguration("Compiler version is not supported, c++17 support is required")
+
+    def configure(self):
+        self.options["websocketpp"].asio = "standalone"
```
Comment (uilianries) [inline recipes/foxglove-websocket/all/conanfile.py:33]: Okay. The default value is "boost", without it, this package would be invalid.

File: recipes/foxglove-websocket/all/test_package/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package CXX)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_executable(test_package test_package.cpp)
+set_target_properties(test_package PROPERTIES CXX_STANDARD 17 CXX_STANDARD_REQUIRED ON)
+target_link_libraries(test_package ${CONAN_LIBS})
```
Comment (uilianries) [inline recipes/foxglove-websocket/all/test_package/CMakeLists.txt:9]:
```suggestion
conan_basic_setup(TARGETS)

find_package(foxglove-websocket REQUIRED CONFIG)

add_executable(test_package test_package.cpp)
set_target_properties(test_package PROPERTIES CXX_STANDARD 17 CXX_STANDARD_REQUIRED ON)
target_link_libraries(test_package foxglove-websocket::foxglove-websocket)
```

File: recipes/foxglove-websocket/all/test_package/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
+from conans import ConanFile, CMake, tools
+import os
+
+
+class TestPackageConan(ConanFile):
+    settings = ("os", "arch", "compiler", "build_type")
+    generators = "cmake"
```
Comment (uilianries) [inline recipes/foxglove-websocket/all/test_package/conanfile.py:7]:
```suggestion
    generators = "cmake", "cmake_find_package"
```

File: recipes/foxglove-websocket/all/test_package/CMakeLists.txt (hunk 5 above)
Comment (jtbandes) [reply recipes/foxglove-websocket/all/test_package/CMakeLists.txt]: It seems like that didn't work.

File: recipes/foxglove-websocket/all/test_package/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+from conans import ConanFile, CMake, tools
+import os
+
+
+class TestPackageConan(ConanFile):
+    settings = ("os", "arch", "compiler", "build_type")
+    generators = ("cmake", "cmake_find_package")
```
Comment (uilianries) [inline recipes/foxglove-websocket/all/test_package/conanfile.py:7]:
```suggestion
    generators = ("cmake", "cmake_find_package_multi")
```
it generates cmake config file:
https://docs.conan.io/en/latest/integrations/build_system/cmake/cmake_find_package_multi_generator.html

**Review by uilianries — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**